### PR TITLE
Infrastructure additions

### PR DESCRIPTION
**Pull request by sethrj**
- Random number helper function and such
- Vector utilities
- Add cuda macros to SoftEqual

### CODE REVIEW

File: src/random/distributions/GenerateCanonical.i.hh [hunk 1]
```diff
@@ -12,35 +12,34 @@ namespace celeritas
...
 
-#ifdef __NVCC__
 //---------------------------------------------------------------------------//
 /*!
- * Specialization for RngEngine, float
+ * Helper function to generate a random real number in [0, 1).
  */
-__device__ float GenerateCanonical<RngEngine, float>::operator()(RngEngine& rng)
+template<class RealType, class Generator>
+CELER_FUNCTION RealType sample_uniform(Generator& g)
```
Comment (amandalund) [inline src/random/distributions/GenerateCanonical.i.hh:29]:
```suggestion
CELER_FUNCTION RealType generate_canonical(Generator& g)
```

File: src/random/distributions/GenerateCanonical.i.hh [hunk 2]
```diff
@@ -12,35 +12,34 @@ namespace celeritas
...
 {
-    return curand_uniform(rng.state_);
+    return GenerateCanonical<Generator, RealType>()();
```
Comment (amandalund) [inline src/random/distributions/GenerateCanonical.i.hh:31]:
```suggestion
    return GenerateCanonical<Generator, RealType>()(g);
```

File: src/random/distributions/GenerateCanonical.i.hh [hunk 3]
```diff
@@ -12,35 +12,34 @@ namespace celeritas
...
 
 //---------------------------------------------------------------------------//
 /*!
- * Specialization for RngEngine, double
+ * Helper function to generate a random real number in [0, 1).
  */
-__device__ double
-GenerateCanonical<RngEngine, double>::operator()(RngEngine& rng)
+template<class Generator>
+CELER_FUNCTION real_type sample_uniform(Generator& g)
```
Comment (amandalund) [inline src/random/distributions/GenerateCanonical.i.hh:39]:
```suggestion
CELER_FUNCTION real_type generate_canonical(Generator& g)
```

File: src/random/distributions/GenerateCanonical.i.hh [hunk 4]
```diff
@@ -12,35 +12,34 @@ namespace celeritas
...
 {
-    return curand_uniform_double(rng.state_);
+    return GenerateCanonical<Generator, real_type>()();
```
Comment (amandalund) [inline src/random/distributions/GenerateCanonical.i.hh:41]:
```suggestion
    return GenerateCanonical<Generator, real_type>()(g);
```

File: src/random/distributions/BinaryDistribution.hh [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+    {
+        // ...
+    }
+    BinaryDistribution also_snake_eyes(1.0 / 36.0);
+   \endcode
+ *
+ * \note Since `true` is 1 and `false` is 0, the constructor argument is the
+ * _opposite_ order of std::discrete_distribution.
+ */
+class BinaryDistribution
```
Comment (amandalund) [inline src/random/distributions/BinaryDistribution.hh:33]: Thoughts on naming this `BernoulliDistribution`?

File: src/random/distributions/GenerateCanonical.i.hh (hunk 1 above)
Comment (sethrj) [reply src/random/distributions/GenerateCanonical.i.hh:29]: Uh oh, if this is still here how is everything building and passing? 😱 Good catch!

File: src/random/distributions/BinaryDistribution.hh (hunk 5 above)
Comment (sethrj) [reply src/random/distributions/BinaryDistribution.hh:33]: Wow I'm a dum-dum, I didn't even know of that distribution's existence (or that a binary distribution is called a Bernoulli distribution). That's exactly what this should be. I'll update. Thanks!

File: src/random/distributions/GenerateCanonical.i.hh (hunk 4 above)
Comment (sethrj) [reply src/random/distributions/GenerateCanonical.i.hh:41]: Clearly I must be missing a test if the code is building as-iss...

Comment (amandalund) [reply src/random/distributions/GenerateCanonical.i.hh:41]: Haha yep, just those two helper functions are never used/tested.

**Comment by sethrj** [comment]
OK @amandalund , everything's fixed now -- I changed all the distributions to use `generate_canonical` so that `GenerateCanonical` becomes an implementation detail (good for overriding with the various RNGs). I also updated the name of the distribution as you suggested. Finally, I realized I wasn't ever building with warnings on, so I changed my build flags to add warnings and fixed a couple of warnings that showed up.

**Comment by sethrj** [comment]
Thanks @amandalund !

**Pull request merged**